File: src/main/java/com/practice/mediumalgo/Subset.java
package com.practice.mediumalgo;

import java.util.ArrayList;
import java.util.List;

public class Subset {
    public static void main(String[] args){

        Subset subset = new Subset();
        subset.subsets(new int[]{1,2,3});

    }

    public List<List<Integer>> subsets(int[] nums) {
        List<List<Integer>> output = new ArrayList();
        output.add(new ArrayList<Integer>());



        for(int num: nums){
            List<List<Integer>> newSubsets = new ArrayList();
            for (List<Integer> curr : output){
                List<Integer> a = new ArrayList<Integer>(curr);
                a.add(num);
                newSubsets.add(a);
            }
            for (List<Integer> curr : newSubsets) {
                output.add(curr);
            }
        }
        return output;
    }
}

File: src/main/scala/com/practice/med/Subset.scala
package com.practice.med

object Subset {

  def main(args: Array[String]): Unit = {
    println(subsets(Array(1, 2, 3)))
  }

  def subsets(nums: Array[Int]): List[List[Int]] = {
    nums.foldLeft(List(List.empty[Int])) {
      (acc, num) => { acc ++ acc.map { f: List[Int] => f :+ num }
      }
    }
  }
}
